src/wasm/wasm-validator.cpp

```diff
@@ -2779,6 +2780,36 @@ void FunctionValidator::visitFunction(Function* curr) {
     Name name = pair.second;
     shouldBeTrue(seen.insert(name).second, name, "local names must be unique");
   }
+
+  if (getModule()->features.hasGCNNLocals()) {
+    // If we have non-nullable locals, verify that no local.get can read a null
+    // default value.
+    // TODO: this can be fairly slow due to the LocalGraph. writing more code to
+    //       do a more specific analysis (we don't need to know all sets, just
+    //       if there is a set of a null default value that is read) could be a
+    //       lot faster.
+    bool hasNNLocals = false;
+    for (const auto& var : curr->vars) {
+      if (var.isNonNullable()) {
+        hasNNLocals = true;
+        break;
+      }
+    }
+    if (hasNNLocals) {
+      LocalGraph graph(curr);
+      for (auto& [get, sets] : graph.getSetses) {
+        if (curr->getLocalType(get->index).isNonNullable()) {
+          for (auto* set : sets) {
+            // If this is a null, then it is the value of a param or the default
+            // value of a var. The latter case is an error.
+            shouldBeTrue(!!set || curr->isParam(get->index),
```

Comment:
We could hoist the check for whether this local is a param out of this loop to avoid some unnecessary work.

Reply:
Good idea, simplified now.